### PR TITLE
[Feature] Add tool for converting labelme data to coco format

### PR DESCRIPTION
**Pull request by lllrj**
## Motivation
Provides a tool for labelme2coco that does not include the ability to assemble initially scattered labelme files into a single file.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: tools/misc/labelme2coco.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+import glob
+import json
+import os
+import shutil
+
+import numpy as np
+from sklearn.model_selection import train_test_split
+from tqdm import tqdm
+
+# import sys
```
Comment (jin-s13) [inline tools/misc/labelme2coco.py:12]: please remove unnecessary codes.

File: tools/misc/labelme2coco.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+
+class Labelme2coco_keypoints():
+
+    def __init__(self, args):
+        """Lableme 关键点数据集转 COCO 数据集的构造函数:
```
Comment (jin-s13) [inline tools/misc/labelme2coco.py:18]: How about translating these sentences into English?

File: tools/misc/labelme2coco.py (hunk 1 above)
Comment (lllrj) [reply tools/misc/labelme2coco.py:12]: sorry,I think some commented code might be useful so kept it.

File: tools/misc/labelme2coco.py (hunk 2 above)
Comment (lllrj) [reply tools/misc/labelme2coco.py:18]: I don't have much time these days because of my studies,could you please help me?

File: tools/misc/labelme2coco.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        """
+
+        for name, id in self.classname_to_id.items():
+            category = {}
+
+            category['supercategory'] = name
+            category['id'] = id
+            category['name'] = name
+            # n个关键点数据
+            category['keypoint'] = [
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:191]: Should be another input instead of hard-coded.

File: tools/misc/labelme2coco.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                         ]  # 每个keypoint数组初始化为[0,..] len = 36 对应12个点(x,y,v)
+            num_keypoints = 0
+
+            label = object['label']
+            bbox = object['points']
+            annotation['id'] = self.ann_id
+            annotation['image_id'] = self.img_id
+            annotation['category_id'] = int(self.classname_to_id[label])
+            annotation['iscrowd'] = 0
+            annotation['area'] = 1.0
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:145]: `area` should be calculated, not 1.0

Comment (lllrj) [reply tools/misc/labelme2coco.py:145]: oh I forgot to modify this code when I found it,thank you!

File: tools/misc/labelme2coco.py (hunk 2 above)
Comment (LareinaM) [inline tools/misc/labelme2coco.py:18]:
```suggestion
        """Constructor for Lableme keypoint dataset to COCO dataset.
```

File: tools/misc/labelme2coco.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+                                          ]  # 两个坐标点
+            annotation['bbox'] = self._get_box(bbox)  # 矩形框左上角的坐标和矩形框的长宽
+
+            # 生成keypoint的list
+            for keypoint in keypoints_list:
+                point = keypoint['points']
+                label = keypoint['label']  # 点的名字
+                num_keypoints = self._get_keypoints(point[0], keypoints,
+                                                    num_keypoints, label)
+            annotation['keypoints'] = keypoints
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:156]: Your `keypoints ` never changes

File: tools/misc/labelme2coco.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+                1,                  # 相当于 Z 值，2D关键点 v = 0表示不可见,
+                                      v = 1表示标记但不可见，v = 2表示标记且可见
+                82.42582269256718,
+                109.95672933232304,
+                1,
+                ...,
+            ],
+        """
+        labels = ['wrist', 'thumb1', 'thumb2', ...]
+        flag = label.split('_')[-1]
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:69]: There is a `flag` entry in Labelme format, consider using that?

File: tools/misc/labelme2coco.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+                122.5043507571318,  # y 的值
+                1,                  # 相当于 Z 值，2D关键点 v = 0表示不可见,
+                                      v = 1表示标记但不可见，v = 2表示标记且可见
+                82.42582269256718,
+                109.95672933232304,
+                1,
+                ...,
+            ],
+        """
+        labels = ['wrist', 'thumb1', 'thumb2', ...]
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:68]: Should not be hard-coded.

File: tools/misc/labelme2coco.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        x = label.split('_')[0]
+        visible = 0
+        if flag == 'occluded':
+            visible = 1
+        else:
+            visible = 2
+        x = labels.index(x)
+        keypoints[x * 3] = points[0]
+        keypoints[x * 3 + 1] = points[1]
+        keypoints[x * 3 + 2] = visible
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:79]: `keypoints` modified but not returned.

File: tools/misc/labelme2coco.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+    def __init__(self, args):
+        """Lableme 关键点数据集转 COCO 数据集的构造函数:
+
+        Args
+            args：命令行输入的参数
+                - class_name 根类名字
+        """
+
+        self.classname_to_id = {args.class_name: 1}
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:25]: Is it possible to have multiple classes?

File: tools/misc/labelme2coco.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+                elif shape['shape_type'] == 'point':  # keypoints
+                    keypoints_list.append(shape)
+            # 输入为一个文件的keypoints和bbox，即一张图里的信息
+            self._annotation(bboxes_list, keypoints_list, json_path)
+
+        keypoints = {}
+        keypoints['info'] = {
+            'description': 'Air Dataset',
+            'version': 1.0,
+            'year': 2022
```
Comment (LareinaM) [inline tools/misc/labelme2coco.py:227]:
```suggestion
            'year': datetime.datetime.now().year
```

File: tools/misc/labelme2coco.py (hunk 5 above)
Comment (lllrj) [reply tools/misc/labelme2coco.py:156]: it is changed in `self._get_keypoints()`

Comment (LareinaM) [reply tools/misc/labelme2coco.py:156]: > it is changed in `self._get_keypoints()`

In your `self._get_keypoints()`, `keypoints` is passed by value, modified but never returned.

File: tools/misc/labelme2coco.py (hunk 6 above)
Comment (lllrj) [reply tools/misc/labelme2coco.py:69]: About this `flag`, I use it to distinguish from occluded or unoccluded keypoints because of the coco format 'keypoint = [x,y,v,....]'.
this v means visible for occluding

Comment (LareinaM) [reply tools/misc/labelme2coco.py:69]: > About this `flag`, I use it to distinguish from occluded or unoccluded keypoints because of the coco format 'keypoint = [x,y,v,....]'. this v means visible for occluding

Each point in Labelme format has a `flags` field for you to add flags, and you can use it to distinguish the occluded keypoints, it is better than writing all flags in the label.

Comment (lllrj) [reply tools/misc/labelme2coco.py:69]: > > About this `flag`, I use it to distinguish from occluded or unoccluded keypoints because of the coco format 'keypoint = [x,y,v,....]'. this v means visible for occluding
> 
> Each point in Labelme format has a `flags` field for you to add flags, and you can use it to distinguish the occluded keypoints, it is better than writing all flags in the label.

I didn't understand the role of the flag before,thanks.

File: tools/misc/labelme2coco.py (hunk 5 above)
Comment (lllrj) [reply tools/misc/labelme2coco.py:156]: I think it might be passed by address?it is defined in line 135.

**Review by LareinaM — COMMENTED** [review]
Re-formatted the codes and add a corresponding document.
